scripts/repository.py

```diff
@@ -0,0 +1,229 @@
+
+        if not config_url or not config_url.lower() in [url, url+".git"]:
+            continue
+        if not config_path:
+            logging.debug("Found submodule config = %s", config)
+            raise UserWarning(f"Can't find path to submodule '{url}'")
+        return Path(repo/config_path if abspath else config_path)
+
+    logging.debug("Found submodules = %s", submodules)
+    raise UserWarning(f"Can't find submodule '{url}'")
+
+def litani_root(submodules=None, repo=None, abspath=True):
+    """Root of litani submodule."""
+
+    litani = 'https://github.com/awslabs/aws-build-accumulator'
+    return submodule_root(litani, submodules, repo, abspath)
+
+def starter_kit_root(submodules=None, repo=None, abspath=True):
+    """Root of starter kit submodule."""
+
+    old_starter = 'https://github.com/awslabs/aws-templates-for-cbmc-proofs'
+    new_starter = 'https://github.com/model-checking/cbmc-starter-kit'
+    return (submodule_root(old_starter, submodules, repo, abspath) or
+            submodule_root(new_starter, submodules, repo, abspath))
+
+################################################################
+# Discover the set of all source files in the repository that define a
+# function named func.
+
+def repo_sources(repo='.'):
+    """Paths to all source files within a repository.
+
+    Return a list of paths to C source files found under the
+    repository root (paths are relative to the repository root)."""
+
+    repo = Path(repo).resolve()
+    try:
+        cmd = ['find', '.', '-name', '*.c']
+        kwds = {'cwd': repo, 'capture_output': True, 'text': True}
+        return subprocess.run(cmd, **kwds, check=True).stdout.splitlines()
+    except subprocess.CalledProcessError as error:
+        logging.debug("Failed to run find in %s", repo)
+        logging.debug(error)
+        return []
+
+def function_tags(sources, repo='.'):
+    """Tags for all function definitions in source files.
+
+    Return a list of function definitions found in a list of source
+    files.  Restrict source files to existing files under the root repo.
+    Describe each function definition with a dict
+      {"name": function_name, "path": source_path}
+    giving the function name and the path to the defining source file."""
+
+    def sources_under_root(sources, repo):
+        """Restrict sources to paths under repo relative to repo."""
+
+        srcs = []
+        for src in sources:
+            path = (repo / src).resolve()
+            try:
+                if not path.is_file():
+                    raise ValueError
+                srcs.append(path.relative_to(repo))
+            except ValueError: # path is not a file or not relative to repo
+                logging.debug("Skipping %s: not a file under %s", src, repo)
+                continue
+        return [str(src) for src in srcs]
+
+    repo = Path(repo).resolve()
+    sources = sources_under_root(sources, repo)
+    try:
+        cmd = [
```

Comment:
Creating a command line this long isn't going to work in general, with a few thousand C files you might easily have a command line longer than ARG_MAX.

Suggest to create a `Popen` object with the options you have below plus `stdin=subprocess.PIPE`, and also pass `"-L", "-"` to `ctags`. Then you can use `communicate()` to send the newline-delimited list of C files to ctags' stdin. Again, the output is probably going to be too big here for `run` anyway.

More generally, is there a reason why you don't just use the ctags `--recurse` option?

Reply:
I understand, but this method is VASTLY faster.  I don't understand why.

Reply:
Done.  TL;DR: I was using the ctags option --filter=yes and should have been using -L - (the --help documentation made them sound the same).